src/server/pfs/server/driver.go

```diff
@@ -1809,6 +1283,201 @@ func (d *driver) clearCommit(ctx context.Context, commit *pfs.Commit) error {
+	sort.Slice(toUpdate, func(i, j int) bool { return len(toUpdate[i].Provenance) < len(toUpdate[j].Provenance) })
+	// re-compute the complete provenance of branchInfo and all its subvenant branches
+	for _, bi := range toUpdate {
+		bi.Provenance = make([]*pfs.Branch, 0)
+		for _, directProv := range bi.DirectProvenance {
+			add(&bi.Provenance, directProv)
+			directProvBI, err := getBranchInfo(directProv)
+			if err != nil {
+				return err
+			}
+			for _, p := range directProvBI.Provenance {
+				add(&bi.Provenance, p)
+			}
+		}
+	}
+	// add branchInfo and its subvenance to the subvenance of all of its provenance branches
+	for _, p := range branchInfo.Provenance {
+		pbi, err := getBranchInfo(p)
+		if err != nil {
+			return err
+		}
+		for _, ubi := range toUpdate {
+			add(&pbi.Subvenance, ubi.Branch)
+		}
+	}
+	// remove branchInfo and its subvenance from all branches that are no longer in branchInfo's Provenance
+	oldProvenance := make([]*pfs.Branch, 0)
+	for _, odp := range oldDirectProvenance {
+		if !has(&branchInfo.Provenance, odp) {
+			add(&oldProvenance, odp)
+			oldProvenance = append(oldProvenance, odp)
+			opbi, err := getBranchInfo(odp)
+			if err != nil {
+				return err
+			}
+			for _, b := range opbi.Provenance {
+				add(&oldProvenance, b)
+			}
+		}
+	}
+	for _, op := range oldProvenance {
+		opbi, err := getBranchInfo(op)
+		if err != nil {
+			return err
+		}
+		for _, bi := range toUpdate {
+			if !has(&bi.Provenance, op) {
+				del(&opbi.Subvenance, bi.Branch)
+			}
+		}
+	}
+	// now that all Provenance + Subvenance fields are up to date, save all the branches
+	for _, updateBi := range branchInfoCache {
+		if err := d.branches.ReadWrite(txnCtx.SqlTx).Put(updateBi.Branch, updateBi); err != nil {
+			return errors.EnsureStack(err)
+		}
+	}
+	return nil
+}
+
+// for a DAG to be valid, it may not have a multiple branches from the same repo
+// reachable by traveling edges bidirectionally. The reason is that this would complicate resolving
+func (d *driver) validateDAGStructure(txnCtx *txncontext.TransactionContext, bi *pfs.BranchInfo) error {
+	branchInfoCache := map[string]*pfs.BranchInfo{pfsdb.BranchKey(bi.Branch): bi}
+	getBranchInfo := func(b *pfs.Branch) (*pfs.BranchInfo, error) {
+		if bi, ok := branchInfoCache[pfsdb.BranchKey(b)]; ok {
+			return bi, nil
+		}
+		bi := &pfs.BranchInfo{}
+		if err := d.branches.ReadWrite(txnCtx.SqlTx).Get(b, bi); err != nil {
+			return nil, errors.Wrapf(err, "get branch info")
+		}
+		branchInfoCache[pfsdb.BranchKey(b)] = bi
+		return bi, nil
+	}
+	expanded := make(map[string]struct{}) // expanded branches
+	for {
+		lastExpansion := len(branchInfoCache)
+		bis := make([]*pfs.BranchInfo, 0)
+		for _, bi := range branchInfoCache {
+			bis = append(bis, bi)
+		}
+		for _, bi := range bis {
+			if _, ok := expanded[bi.Branch.String()]; ok {
+				continue
+			}
+			expanded[bi.Branch.String()] = struct{}{}
+			for _, b := range append(bi.Provenance, bi.Subvenance...) {
```

Comment:
this is not good. Iterate over prov & subv separately. append could modify bi.Provenance if there's remaining capacity